js/ui/collection/ui.collection_widget.base.js

```diff
@@ -603,19 +669,86 @@ var CollectionWidget = Widget.inherit({
         this._startIndexForAppendedItems = null;
     },
 
-    _dataSourceChangedHandler: function(newItems) {
-        var items = this.option("items");
-        if(this._initialized && items && this._shouldAppendItems()) {
-            this._renderedItemsCount = items.length;
-            if(!this._isLastPage() || this._startIndexForAppendedItems !== -1) {
-                this.option().items = items.concat(newItems.slice(this._startIndexForAppendedItems));
+    _deleteItemElement: function($item, deletedActionArgs) {
+        $item.remove();
+        this._fireDeleted($item, deletedActionArgs);
+    },
+
+    _getWrapperElement: commonUtils.noop,
+
+    _modifyByChanges: function(items, changes) {
```

Comment:
This method is too big. Extract code from update, insert. delete cases to separate method